wallet: Remove calls to Chain::Lock methods

ryanofsky opened the pull request:
This is a set of changes updating wallet code to make fewer calls to `Chain::Lock` methods, so the `Chain::Lock` class will be easier to remove in #16426 with fewer code changes and small changes to behavior.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in tryGetTxStatus and tryGetBalances

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

It also helps ensure the GUI display stays up to date in the case where the
node chain height runs ahead of wallet last block processed height.

ryanofsky committed:
wallet refactor: Avoid use of Chain::Lock in qt wallettests

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change doesn't affect behavior.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in importprunedfunds

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip, in which case the "Block not found in chain" error
will be stricter and not allow importing data from a blocks between the wallet
last processed tip and the current node tip.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in importwallet and dumpwallet

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip, in which case it will use more accurate backup and
rescan timestamps.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in importmulti

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip, in which case it may use a more accurate rescan
time.

jnewbery commented:
Concept ACK

ryanofsky committed:
wallet: Avoid use of Chain::Lock in listsinceblock

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip. Previously listsinceblock might not have returned
all transactions up to the claimed "lastblock" value in this case, resulting in
race conditions and potentially missing transactions in cases where
listsinceblock was called in a loop like
https://github.com/bitcoin/bitcoin/issues/14338#issuecomment-426706574

ariard commented on src/interfaces/wallet.cpp:
```diff
@@ -330,13 +330,9 @@ class WalletImpl : public Wallet
-        }
+        num_blocks = m_wallet->GetLastBlockHeight();
+        block_time = -1;
+        m_wallet->chain().findBlock(m_wallet->GetLastBlockHash(), nullptr, &block_time);
```
cfc9373

As noted in commit, we may have asynchronicity between node and wallet w.r.t block processing. So returned block time may not be the one of last block height. To avoid this we may cache block time to fetch it when needed with `GetLastBlockTime`. It would also remove some `getBlockTime` (but not all last time I looked on).

ariard replied:
Even further than cs_wallet, we are still holding cs_main there through the chain lock. When we're going to remove locked_chain we may have asynchronicity due to to height being based on BlockConnected locked by cs_wallet and findBlock locked by cs_main ?

ariard commented on src/wallet/rpcdump.cpp:
```diff
@@ -359,8 +359,9 @@ UniValue importprunedfunds(const JSONRPCRequest& request)
-    if (height == nullopt) {
+    LOCK(pwallet->cs_wallet);
+    int height;
+    if (!pwallet->chain().findAncestor(pwallet->GetLastBlockHash(), merkleBlock.header.GetHash(), &height)) {
```
e399fb4

Previously, `getBlockHeight` would have return nullopt if `merkleBlock` have been out of chain. With this change, a height can be returned and ancestry asserted while node and walle tip being unsynchronized, so `merkleBlock` have been reorged out. IMO that's fine if rpc caller is aware than processing have been done with best-wallet-knowledge.

ariard commented on src/wallet/rpcdump.cpp:
```diff
@@ -789,9 +788,10 @@ UniValue dumpwallet(const JSONRPCRequest& request)
-    file << strprintf("#   mined on %s\n", tip_height ? FormatISO8601DateTime(locked_chain->getBlockTime(*tip_height)) : "(missing block time)");
+    file << strprintf("# * Best block at time of backup was %i (%s),\n", pwallet->GetLastBlockHeight(), pwallet->GetLastBlockHash().ToString());
+    int64_t block_time = 0;
+    pwallet->chain().findBlock(pwallet->GetLastBlockHash(), nullptr, &block_time);
```
673e0b6

Another candidate for `GetLastBlockTime`

ariard commented on src/wallet/rpcdump.cpp:
```diff
@@ -564,8 +564,7 @@ UniValue importwallet(const JSONRPCRequest& request)
         }
-        Optional<int> tip_height = locked_chain->getHeight();
-        nTimeBegin = tip_height ? locked_chain->getBlockTime(*tip_height) : 0;
+        pwallet->chain().findBlock(pwallet->GetLastBlockHash(), nullptr, &nTimeBegin);
```
673e0b6

I'm not sure about the commit message, IMO it's less accurate but on the whole make the rescan protection better by starting farther in the past. Anyway, being based on wallet tip or node tip should be safe given the range of TIMESTAMP_WINDOW

ryanofsky replied:
> [673e0b6](https://github.com/bitcoin/bitcoin/commit/673e0b6e7c9f671096f4056f9db69ec43d5b4a3b)
> 
> I'm not sure about the commit message, IMO it's less accurate but on the whole make the rescan protection better by starting farther in the past. Anyway, being based on wallet tip or node tip should be safe given the range of TIMESTAMP_WINDOW

Hmm, I'm not sure when it would be less accurate. Are you thinking of a case?

ariard replied:
I think I meaned because we rely now on wallet last block hash instead of main tip and we may be late from one block, so in my opinion we are _less_ accurate from one block but we agree on rescan being safer.

ryanofsky committed:
wallet refactor: Avoid use of Chain::Lock in CWallet::RescanFromTime

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change has no effect on behavior.

ariard commented on src/wallet/rpcdump.cpp:
```diff
@@ -1364,20 +1364,13 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
 
         // Verify all timestamps are present before importing any keys.
-        const Optional<int> tip_height = locked_chain->getHeight();
-        now = tip_height ? locked_chain->getBlockMedianTimePast(*tip_height) : 0;
```
1f4b604

If #17443 gets first + `GetLastBlockTime`, you may avoid to call `findBlock` here.

ariard reviewed (approved):
Code review ACK 1f4b604. 

What do you think about caching also last block time at connection ? I think that would make some changes here better by avoiding to uselessly lock the node, even if it's through the interface.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in rescanblockchain

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip. The rescanblockchain error height error checking
will just be stricter in this case and only accept values up to the last
processed height

ryanofsky reviewed (commented):
Thanks for the review! I pushed a few more commits I think finishing up the wallet rpc code. There are some more changes to come in wallet.cpp after this.

Updated 1f4b60471730fb5961096fef286c1bd6ec2538b8 -> aeba8afaf5d12fd7b2552616f817712d4fc5062e ([`pr/unlock.1`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.1) -> [`pr/unlock.2`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.1..pr/unlock.2)) with some tweaks to earlier commits and a few new commits extending the PR

re: https://github.com/bitcoin/bitcoin/pull/17954#pullrequestreview-346148268

> What do you think about caching also last block time at connection ? I think that would make some changes here better by avoiding to uselessly lock the node, even if it's through the interface.

I think it's probably a good idea. It would use some more memory though, and it's unclear if it the change would simplify this PR or not overlap much. I think it's probably something to try out separately.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in CWallet::ScanForWalletTransactions

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change affects behavior in a few small ways.

- If there's no max_height specified, percentage progress is measured ending at
  wallet last processed block instead of node tip

- More consistent error reporting: Early check to see if start_block is on the
  active chain is removed, so start_block is always read and the triggers an
  error if it's unavailable

ryanofsky committed:
wallet: Avoid use of Chain::Lock in CWallet::CreateTransaction

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip, where it may set a different lock time.

ryanofsky committed:
wallet: Avoid use of Chain::Lock in CWallet::GetKeyBirthTimes

This is a step toward removing the Chain::Lock class and reducing cs_main
locking.

This change only affects behavior in the case where wallet last block processed
falls behind the chain tip, where it will treat the last block processed as the
current tip.

ryanofsky commented:
Added 3 commits aeba8afaf5d12fd7b2552616f817712d4fc5062e -> 60e6595f5cfd266c14c48994b0bb4afb5bf7fcb3 ([`pr/unlock.2`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.2) -> [`pr/unlock.3`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.2...pr/unlock.3)) and removed PR draft status. 

I think this is basically done. Combined with #17443 it should remove all calls to `interfaces::Chain::Lock` methods after wallet loading (#15719 should clean up loading). I'm hoping this PR and #17443 can be merged before #16426 so #16426 can be a little smaller and change wallet behavior less.

ariard commented:
Thanks Russ, will review new changes soon.

> I think it's probably a good idea. It would use some more memory though, and it's unclear if it the change would simplify this PR or not overlap much. I think it's probably something to try out separately.

I've a branch doing it, I can try to rebase it on top of this one and squeeze a last one before #16426.

ryanofsky reviewed (commented):
Comments below are from last week (they were in pending status because I forgot to submit the github review)

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3546,22 +3545,23 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
         if (!request.params[0].isNull()) {
             start_height = request.params[0].get_int();
-            if (start_height < 0 || !tip_height || start_height > *tip_height) {
+            if (start_height < 0 || tip_height < 0 || start_height > tip_height) {
```
`GetLastBlockHeight()` can't return a `tip_height` that's < 0, so I think you can just remove `|| tip_height < 0 `

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3546,22 +3545,23 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
     {
         auto locked_chain = pwallet->chain().lock();
-        Optional<int> tip_height = locked_chain->getHeight();
+        LOCK(pwallet->cs_wallet);
```
Do you need to hold the wallet lock for this entire block? Does it make sense to call:

```
WITH_LOCK(pwallet->cs_wallet, pwallet->GetLastBlockHeight());
```

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3570,25 +3570,17 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
-            if (stop_height) {
-                stop_block = locked_chain->getBlockHash(*stop_height);
-            }
+        if (tip_height >= 0) {
```
Again, I think this is always true, so you can remove this conditional.

jnewbery reviewed (commented):
I've reviewed as far as 759e49497766e15fa57b332527b800e95b153113 _wallet: Avoid use of Chain::Lock in rescanblockchain_ and just have nits so far. I'll try to complete review tomorrow.

ryanofsky reviewed (commented):
Updated d2f92a9f759d41b150fa702c8f83d1b05b11c943 -> 1125fd92bd3ee6d44ab9f8118ac68c0f621c5a94 ([`pr/unlock.8`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.8) -> [`pr/unlock.9`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.8..pr/unlock.9)) with some hasBlocks improvements and other suggested changes

ariard commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3569,25 +3570,15 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
-                stop_block = locked_chain->getBlockHash(*stop_height);
-            }
-        }
+        start_block = pwallet->chain().findAncestorByHeight(pwallet->GetLastBlockHash(), start_height);
```
9aa4b6b

I find `findAncestorByHeight` unclear, here we may have `start_height` and `GetLastBlockHash` not pointing to same block. Behavior follows method documentation but why bother asking for the hash, query in ChainActive with the provided height ?

Honestly here I would prefer to stick with `getBlockHash`, behavior is more straightforward.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1649,7 +1649,7 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
         block_height = locked_chain->getBlockHeight(block_hash);
         progress_begin = chain().guessVerificationProgress(block_hash);
-        progress_end = chain().guessVerificationProgress(stop_block.IsNull() ? tip_hash : stop_block);
+        progress_end = chain().guessVerificationProgress(max_height ? chain().findAncestorByHeight(tip_hash, *max_height) : tip_hash);
```
9aa4b6b

Same here, why `ScanForWalletTransactions` function to then add a call to get previously furnished information ? I would prefer to keep removed `getBlockHash` calls in `rescanblockchain`

ariard reviewed (commented):
Reviewed until 9aa4b6b

I think this approach is better than mine in #16426 by documenting better behavior changes. However I would prefer to avoid introducing to much new method with less-understood behaviors, is the design goal of this patchset still to avoid any reorg-unsafe-method after chain lock is removed ?

ryanofsky reviewed (commented):
Thank you Antoine for detailed review! I tried to answer all your comments below

ryanofsky commented on src/interfaces/chain.h:
```diff
@@ -133,12 +133,23 @@ class Chain
+    //! Find most recent common ancestor between two blocks and optionally
+    //! return its hash and/or height. Also return height of first block. Return
+    //! nullopt if either block is unknown or there is no common ancestor.
+    virtual Optional<int> findCommonAncestor(const uint256& block_hash1,
```
Here's a change that would make all the find block methods return block information same way 6f74c0a042b001283e1d7dd8a8ad8b46c75328e5 ([branch](https://github.com/ryanofsky/bitcoin/commits/pr/getblock)), if it helps

ariard commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1605,8 +1606,7 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
 
-    const Optional<int> tip_height = locked_chain->getHeight();
-    int depth = tip_height && height ? (1 + *tip_height - *height) : -1;
+    int depth = height ? pwallet->GetLastBlockHeight() + 1 - *height : -1;
```
Oh right, it's an Optional, forget about it, forgive my C++ noobiness

ariard commented on src/interfaces/chain.cpp:
```diff
@@ -259,6 +259,16 @@ class ChainImpl : public Chain
+    Optional<uint256> findFirstBlockWithTimeAndHeight(int64_t min_time, int min_height, int* height = nullptr) override
+    {
+        LOCK(::cs_main);
+        CBlockIndex* block = ::ChainActive().FindEarliestAtLeast(min_time, min_height);
```
Hmmm if I understand `RescanFromTime` expected behavior is to find earliest block with _both_ nTime and height superior at the ones passed not _either_ so sounds like I broke it ?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17954#discussion_r379688216

#15670 seems right to me, at least at first glance. `a || b` can only be false if both a and b are false

ryanofsky commented:
Review club notes at https://bitcoincore.reviews/17954.html. Meeting in 2 hours if I'm time zoning correctly

jonatack reviewed (commented):
Concept ACK. Built/ran tests, code review in progress.

luke-jr commented on src/interfaces/chain.h:
```diff
@@ -133,12 +133,23 @@ class Chain
         int* height = 0) = 0;
 
+    //! Find most recent common ancestor between two blocks and optionally
+    //! return its hash and/or height. Also return height of first block. Return
```
I think it's confusing to return something unrelated to the common ancestor here...

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17954#discussion_r381473300

> I think it's confusing to return something unrelated to the common ancestor here...

Agreed, will backport 25c1ae48204215622bc9fd3a8bc9677f15c32674 ([branch](https://github.com/ryanofsky/bitcoin/commits/pr/getblock)) as soon as I get a chance

fjahr reviewed (approved):
Code review ACK 1125fd92bd3ee6d44ab9f8118ac68c0f621c5a94 (also built and ran tests)

Had some nits, this could get merged but I would be happy if you could address my comments on `height` if you still make changes.

ryanofsky committed:
refactor: Add interfaces::FoundBlock class to selectively return block data

FoundBlock class allows interfaces::Chain::findBlock to return more block
information without having lots of optional output parameters. FoundBlock class
is also used by other chain methods in upcoming commits.

There is mostly no change in behavior. Only exception is
CWallet::RescanFromTime now throwing NonFatalCheckError instead of
std::logic_error.

ryanofsky reviewed (commented):
Thanks for reviews and discussion!

Updated 1125fd92bd3ee6d44ab9f8118ac68c0f621c5a94 -> ace85fbe3516ca02d4bbc6e5e0d677f2cf155933 ([`pr/unlock.9`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.9) -> [`pr/unlock.10`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.9..pr/unlock.10)) adding FoundBlock class so find block methods can return information in a uniform way

re: https://github.com/bitcoin/bitcoin/pull/17954#pullrequestreview-362525933

> Had some nits, this could get merged but I would be happy if you could address my comments on `height` if you still make changes.

You had a few comments about heights, and think I resolved or addressed them all, but let me know!

ariard reviewed (commented):
Code review ACK ace85fb minus agreeing on what the behavior changes is https://github.com/bitcoin/bitcoin/pull/17954#discussion_r379172583 (by switching for `findAncestorByHeight` I still think you now return a previous block as `lastblock`)

Other suggestions are leaning using more the new `FoundBlock` helper class instread of adding new methods. I would prefer but I'm fine with how it is right now.

ryanofsky reviewed (commented):
Updated ace85fbe3516ca02d4bbc6e5e0d677f2cf155933 -> 4c7ae7319e5796e4aa7011d26b2dfd6bca4ebe45 ([`pr/unlock.10`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.10) -> [`pr/unlock.11`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.10..pr/unlock.11))

ariard commented on src/interfaces/chain.h:
```diff
@@ -153,6 +153,12 @@ class Chain
 
+    //! Return whether block descends from a specified ancestor, and
+    //! optionally return block information.
+    virtual bool findAncestorByHash(const uint256& block_hash,
```
Okay I get your point with struct-with-accessor-methods vs real-class-with-methods-that-execute-code wrt with IPC/memory separation. My assumption here was we should clean up completely these methods by storing more inside the wallet (like any block header tied to a transaction which matters for us), but that something we should discuss in future PRs/issues. I'm fine with Chain API right now, let's move forward

ariard reviewed (approved):
Code Review ACK 4c7ae73. Changes since last time are more comments, removing `require` method from `FoundBlock`, avoiding short-lived wallet locking.

fjahr commented:
ACK 4c7ae7319e5796e4aa7011d26b2dfd6bca4ebe45

Reviewed code, built and ran tests locally. It took me a moment to get used to the `FoundBlock` abstraction but I agree that it's an improvement.

hebasto commented:
Concept ACK.

hebasto commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3546,22 +3547,23 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
 
     int start_height = 0;
-    uint256 start_block, stop_block;
+    Optional<int> stop_height;
```
g++ compiler `-Wmaybe-uninitialized` warning:
```
wallet/rpcwallet.cpp: In function ‘UniValue rescanblockchain(const JSONRPCRequest&)’:
wallet/rpcwallet.cpp:3550:19: warning: ‘*((void*)& stop_height +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     Optional<int> stop_height;
                   ^~~~~~~~~~~
```

Could be
```suggestion
#include <optional.h>
    Optional<int> stop_height = MakeOptional(false, int());
```
?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17954#discussion_r395695797

> g++ compiler `-Wmaybe-uninitialized` warning:

This is a known false positive with no side effects from an old compiler. https://www.boost.org/doc/libs/1_72_0/libs/optional/doc/html/boost_optional/tutorial/gotchas/false_positive_with__wmaybe_uninitialized.html

I don't think making code less readable to silence these is a good tradeoff. But if silencing them is actually important, we should at least have an automated check, like a linter or an old gcc running on travis and failing so there doesn't have to be a manual reporting, update, and review cycle each time a new instance turns up.

hebasto replied:
> I don't think making code less readable to silence these is a good tradeoff.

Let me add some context: https://github.com/bitcoin/bitcoin/pull/14711#pullrequestreview-193702611, #15292, #18052

> But if silencing them is actually important, we should at least have an automated check, like a linter or an old gcc running on travis and failing so there doesn't have to be a manual reporting, update, and review cycle each time a new instance turns up.

Do you mean adding of the `-Werror=maybe-uninitialized` option to a compiler on Travis?

hebasto commented on src/qt/test/wallettests.cpp:
```diff
@@ -156,7 +156,7 @@ void TestGUI(interfaces::Node& node)
         WalletRescanReserver reserver(wallet.get());
         reserver.reserve();
-        CWallet::ScanResult result = wallet->ScanForWalletTransactions(locked_chain->getBlockHash(0), {} /* stop_block */, reserver, true /* fUpdate */);
+        CWallet::ScanResult result = wallet->ScanForWalletTransactions(Params().GetConsensus().hashGenesisBlock, 0 /* block height */, {} /* max height */, reserver, true /* fUpdate */);
```
Why `auto locked_chain = wallet->chain().lock();` is still needed?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17954#discussion_r395737849

> Why `auto locked_chain = wallet->chain().lock();` is still needed?

Good catch, simplified this code

ryanofsky reviewed (commented):
Thanks for review!

Updated 4c7ae7319e5796e4aa7011d26b2dfd6bca4ebe45 -> 0539e740fb4962d453564adb28e383c77594512d ([`pr/unlock.11`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.11) -> [`pr/unlock.12`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.11..pr/unlock.12)) with suggested test simplification

hebasto commented:
@ryanofsky It seem CI fails due to the conflict with #18234.

hebasto commented on src/interfaces/chain.cpp:
```diff
@@ -256,6 +256,16 @@ class ChainImpl : public Chain
+        if (const CBlockIndex* block = LookupBlockIndex(block_hash)) {
+            if (const CBlockIndex* ancestor = block->GetAncestor(ancestor_height)) {
+                return FillBlock(ancestor, ancestor_out, lock);
+            }
```
nit:
```suggestion
            const CBlockIndex* ancestor = block->GetAncestor(ancestor_height);
            return FillBlock(ancestor, ancestor_out, lock);
```

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-601871454

> @ryanofsky It seem CI fails due to the conflict between [80468a9](https://github.com/bitcoin/bitcoin/commit/80468a97cba27faa9297b86eb901221701d8b13b) and #18234.

Thanks! Rebased 0539e740fb4962d453564adb28e383c77594512d -> 40a8796ad1a4e6032c466c38b8887e8f1e1022a4 ([`pr/unlock.12`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.12) -> [`pr/unlock.13`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.12-rebase..pr/unlock.13)) to fix silent merge conflict with #18234

hebasto commented on src/interfaces/chain.cpp:
```diff
@@ -290,6 +276,17 @@ class ChainImpl : public Chain
+        if (CBlockIndex* block = LookupBlockIndex(block_hash)) {
+            if (max_height && block->nHeight >= *max_height) block = block->GetAncestor(*max_height);
+            for (; block->nStatus & BLOCK_HAVE_DATA; block = block->pprev) {
+                if (block->nHeight < min_height || !block->pprev) return true;
```
Why the second check `!block->pprev` is needed?

hebasto commented on src/interfaces/chain.h:
```diff
@@ -148,6 +148,11 @@ class Chain
+    //! Find next block if block is part of current chain. Also flag if
+    //! there was a reorg and the specified block hash is no longer in the
+    //! current chain, and optionally return block information.
+    virtual bool findNextBlock(const uint256& block_hash, int block_height, const FoundBlock& next={}, bool* reorg=nullptr) = 0;
```
nit: parameter names in the function declaration differ from ones in the function definition:
- `FoundBlock& next` vs `FoundBlock& block_out`
- `bool* reorg` vs `bool* reorg_out`

hebasto commented on src/wallet/wallet.cpp:
```diff
@@ -1619,6 +1619,8 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
  * @param[in] stop_block  Scan ending block. If block is not on the active
  *                        chain, the scan will continue until it reaches the
  *                        chain tip.
```
Is this comment still relevant? And the mention of `stop_block` in `@pre` comment?

hebasto commented on src/wallet/wallet.cpp:
```diff
@@ -2737,17 +2737,19 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
         auto locked_chain = chain().lock();
         LOCK(cs_wallet);
+        tip_hash = GetLastBlockHash();
+        tip_height = GetLastBlockHeight();
```
`tip_hash` and `tip_height` could be `const`:
```suggestion
    CMutableTransaction txNew;
    FeeCalculation feeCalc;
    CAmount nFeeNeeded;
    int nBytes;
    {
        std::set<CInputCoin> setCoins;
        auto locked_chain = chain().lock();
        LOCK(cs_wallet);
        const uint256 tip_hash = GetLastBlockHash();
        const int tip_height = GetLastBlockHeight();
```

hebasto reviewed (approved):
ACK 40a8796ad1a4e6032c466c38b8887e8f1e1022a4, modulo nits, tested on Linux Mint 19.3.

ariard commented:
Code Review ACK 40a8796, changes since last ACK is removing a useless lock tacking in qt test.

(stale Travis?)

ryanofsky reviewed (commented):
Thanks for review and reack!

Rebased 40a8796ad1a4e6032c466c38b8887e8f1e1022a4 -> 19e1db77cbc08f451a3508bb113f2f7cf5a13616 ([`pr/unlock.13`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.13) -> [`pr/unlock.14`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.13-rebase..pr/unlock.14)) due to conflict with #18278, also fixed a hasBlocks edge case and added tests, and made some suggested minor code cleanups

ariard commented:
Code Review ACK 19e1db7

Changes since last time are better documentation, `hasBlock` fix for the lower bound, `findNextBlock` internal simplification.

I agree with @hebasto, `block->pprev` is quite confusing, would be happy to reack a `hasBlock` documentation change to lay out this.

ryanofsky commented:
Thanks for review!

Updated 19e1db77cbc08f451a3508bb113f2f7cf5a13616 -> cdea18ae2dad4a198df65d0043e10c45a22994e3 ([`pr/unlock.14`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.14) -> [`pr/unlock.15`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.14..pr/unlock.15)) adding suggested comment

re: https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-603622770

> I agree with @hebasto, `block->pprev` is quite confusing, would be happy to reack a `hasBlock` documentation change to lay out this.

Added comments, but I am surprised this null check is surprising, range cases are exhaustively checked in the unit test and this just makes results stable and not crashy.

ariard commented:
Code Review ACK cdea18a

What confusing here isn't the null check but the decision to return true in this case, because you may have a void block is the given ranged and still return true.

hebasto commented:
re-ACK cdea18ae2dad4a198df65d0043e10c45a22994e3

promag commented:
Core review ACK cdea18ae2dad4a198df65d0043e10c45a22994e3.

ryanofsky commented:
Rebased cdea18ae2dad4a198df65d0043e10c45a22994e3 -> 710b077f9c0d57a40073ceb9e5568ce37d8cdfe2 ([`pr/unlock.15`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.15) -> [`pr/unlock.16`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.15-rebase..pr/unlock.16)) due to conflict with #18160

hebasto commented:
Reviewed recent [rebase](https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-606702994):

in `interfaces/wallet.cpp` #18160 moved `if (!force && num_blocks == cached_num_blocks) return false;` up, but in f6da44ccce4cfff53433e665305a6fe0a01364e4 it moved back. Could this break behavior introduced by #18160?

ryanofsky commented:
> Reviewed recent [rebase](https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-606702994):
> 
> in `interfaces/wallet.cpp` #18160 moved `if (!force && num_blocks == cached_num_blocks) return false;` up, but in [f6da44c](https://github.com/bitcoin/bitcoin/commit/f6da44ccce4cfff53433e665305a6fe0a01364e4) it moved back. Could this break behavior introduced by #18160?

No, the point of #18160 is to skip recomputing balances when not forced by a transaction update or tip change. Commit f6da44ccce4cfff53433e665305a6fe0a01364e4 is making the tip change detection a little more accurate. An extra lock is required to do this, but GetBalances is still not called. The extra lock is also removed in #16426.

hebasto commented:
re-ACK 710b077f9c0d57a40073ceb9e5568ce37d8cdfe2

ariard commented:
Code Review ACK 710b077

ryanofsky commented:
Current review status:

- [ariard ack](https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-608180917)
- [hebasto ack](https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-607865406)
- [promag pre-rebase ack](https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-605724033)
- [fjahr pre-rebase ack](https://github.com/bitcoin/bitcoin/pull/17954#issuecomment-597674785) 
- [luke review](https://github.com/bitcoin/bitcoin/pull/17954#pullrequestreview-361345920)
- [jonatack review](https://github.com/bitcoin/bitcoin/pull/17954#pullrequestreview-361307966)
- [jnewbery review](https://github.com/bitcoin/bitcoin/pull/17954#pullrequestreview-356971621)

MarcoFalke commented:
> Current review status:

Acks don't help if the changes introduce bugs :red_circle: :dagger: 


ACK 710b077f9c, except the introduced bugs 🤺

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 710b077f9c, except the introduced bugs 🤺
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUitNQv/QXsc/TXCoGbszrOL5HQSrTpPCe9uPGwCUAwd6MCY8ZDmj7JcMRdBeLJ2
iuCPgSMmF1d/GxJbMxTr1J4gVp+dVKkMgneTnPauwgoh537JHEmplmROC8vIixot
RxnW6na6vXfj0aEvSoFbhLBvopXdz5HZVUCjS74/o0YZL2tTvlOlAujHjfJqoFh/
n3XkrCSNtcXyIioJUhgnnH1wFaYPPm8LVgqrgclm03R4kW4hW+Zq6ibLMtZdP+9L
gLWVlMGfDu+C6XGp8Nw/Znd9TgKEIMLY0Yf2FUGk5jMMTlqtoAcjoMVXHYo59QVX
U/YAFl15jnTjoGQsZEuB8Ln1KrvkrkOITS8tPVxjFCGQ8niNY+n1FX+hyA4LDitm
jpUry9hlMswzBP5EFS6SNZ09bho8mHQP/+rnsJ5s3gF2tj+cZN790H8asy+Ml6sG
rtOT/PLGVV7NezB4YSwcW1uJiKDellWOQrA99b/kjwUHkYr01BIILL6XpBMNELfy
4Qv4y0/D
=u2+w
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d8d68feef8d9815d83b13f49567f9148ee13f92ef1f16677d1e9c1c98efcf014  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d8d68feef8d9815d83b13f49567f9148ee13f92ef1f16677d1e9c1c98efcf014f0103452024aa339a93ed75cc1a905cc338b08fff010647cb3242ecc60f413b4d2d3764e668708f1045e90dc5af008f84a96db441942220083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0106f5ffbbd30e43bc209ef5229adaf01d308f1045e90dc5bf0083f6a6fd17340877f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c4183f36e10cf8ae3e437538ca28775b08f1045e90dc5bf008ac14dfeeff14084f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010ff6c1f9254e0a4195f0941a6833af99708f1045e90dc5cf00808200b863f171ed90083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented:
Going to merge, so that the ACKs don't go down the drain

MarcoFalke commented:
actually there are only two acks for the current state of the pr, so I won't merge for now

ryanofsky commented on src/interfaces/chain.h:
```diff
@@ -149,12 +149,24 @@ class Chain
 
+    //! Find ancestor of block at specified height and optionally return
+    //! ancestor information.
+    virtual bool findAncestorByHeight(const uint256& block_hash, int ancestor_height, const FoundBlock& ancestor={}) = 0;
```
re: https://github.com/bitcoin/bitcoin/pull/17954#discussion_r406896059

> [08211e6](https://github.com/bitcoin/bitcoin/commit/08211e640f4432af4a3d67c804ef4a9d0a94edb4)

Updated

ryanofsky reviewed (commented):
Updated 710b077f9c0d57a40073ceb9e5568ce37d8cdfe2 -> 48973402d8bccb673eaeb68b7aa86faa39d3cb8a ([`pr/unlock.16`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.16) -> [`pr/unlock.17`](https://github.com/ryanofsky/bitcoin/commits/pr/unlock.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/unlock.16..pr/unlock.17)) with suggested tweaks and dumpwallet comment fix covered by new test in #18597

MarcoFalke commented:
re-ACK 48973402d8, only change is fixing bug 📀

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 48973402d8, only change is fixing bug 📀
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg7CAwAgLetCNTTZ3j5/d54CEHvjQSQCdU0xAxJZfq2LvcRgphElGFiWjqAnNYM
QCt734Dk5fMXjqj6PtDz7Py0GO0wiinIXbR+bXdVvPY9v251H/zeXkSkDcC+yLr3
2J40jb1Ec/uYMBay7nmXjOmmNJwq4FATKNEBZonLVAqL7z4K7dbJVqyWksVnaR8O
CEUIsEFLjYQ5oTGzuNDL4mLh59lzi/GpSgnManIyLTCn/eEfcDeyTusTIXGnoOG+
6aCYUjLkFTRlkkmUvD3Ife1gGaM5V5/sJ7KsJbrfz5xrrQfJEUN5NXi2m5CL9vJn
0deZmTC3fpbbplW2VJ8gwT7yOwZ4EJPNlvctWFbDapFDov9hSZHb3IlL73XY8d+K
2zpraFjfiiOA/svrngKAd4Jzuj4ZjMHNNNKzLYr0S/0DNmJTfYVYVtGJ+9l/Itl6
uPbFH1au0xdhAgS36e4FtXx1W3CHx60+thQTHe8z1eQo1Jq0R0Y4qhywF2cDfeHC
pR94jjdH
=est8
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `059943689472c2b5db5fc3c5f99636fdd7a6eb8ce233f24227b7d283f6c3d208  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108059943689472c2b5db5fc3c5f99636fdd7a6eb8ce233f24227b7d283f6c3d208f0104ecbf92ac1d6680c592c321ab09c9c0a08fff010227588a3b3e3e70dc03503ca2c667a3c08f1045e9470c1f0080949959e0903259a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108d95b86c500bfa3fe0bebd4d9c95a79108f1045e9470c0f008ff8b31da9ffd03060083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d1097ca21205877267960b7d3d870c1808f1045e9470c1f008b87e94178b9b0f630083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010e70ac66cf6fe75fd7280a01e4810db7108f1045e9470c1f0083ac8d78efe563e260083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

fjahr commented:
re-ACK 48973402d8bccb673eaeb68b7aa86faa39d3cb8a, reviewed rebase and changes since last review, built and ran tests locally

ariard commented:
Coce Review ACK 4897340, only changes are one suggested by last review on more accurate variable naming, human-readable output, args comments in `findCommonAncestor`

practicalswift commented:
@ryanofsky 

Post-merge review comment:

```c++
    bool findCommonAncestor(const uint256& block_hash1, const uint256& block_hash2, const FoundBlock& ancestor_out, const FoundBlock& block1_out, const FoundBlock& block2_out) override
    {
        WAIT_LOCK(cs_main, lock);
        const CBlockIndex* block1 = LookupBlockIndex(block_hash1);
        const CBlockIndex* block2 = LookupBlockIndex(block_hash2);
        const CBlockIndex* ancestor = block1 && block2 ? LastCommonAncestor(block1, block2) : nullptr;
        return FillBlock(ancestor, ancestor_out, lock) & FillBlock(block1, block1_out, lock) & FillBlock(block2, block2_out, lock);
    }
```

Is the use of `&` instead of `&&` intentional on the `return` line?

ryanofsky commented:
> @ryanofsky
> 
> Post-merge review comment:
> 
> ```c++
>     bool findCommonAncestor(const uint256& block_hash1, const uint256& block_hash2, const FoundBlock& ancestor_out, const FoundBlock& block1_out, const FoundBlock& block2_out) override
>     {
>         WAIT_LOCK(cs_main, lock);
>         const CBlockIndex* block1 = LookupBlockIndex(block_hash1);
>         const CBlockIndex* block2 = LookupBlockIndex(block_hash2);
>         const CBlockIndex* ancestor = block1 && block2 ? LastCommonAncestor(block1, block2) : nullptr;
>         return FillBlock(ancestor, ancestor_out, lock) & FillBlock(block1, block1_out, lock) & FillBlock(block2, block2_out, lock);
>     }
> ```
> 
> Is the use of `&` instead of `&&` intentional on the `return` line?

Yes, because `&&` could short circuit and leave block1_out or block2_out data unfilled

practicalswift commented:
@ryanofsky 

Yes, I was thinking something along the lines of:

```c++
    bool findCommonAncestor(const uint256& block_hash1, const uint256& block_hash2, const FoundBlock& ancestor_out, const FoundBlock& block1_out, const FoundBlock& block2_out) override
    {
        WAIT_LOCK(cs_main, lock);
        const CBlockIndex* block1 = LookupBlockIndex(block_hash1);
        const CBlockIndex* block2 = LookupBlockIndex(block_hash2);
        const CBlockIndex* ancestor = block1 && block2 ? LastCommonAncestor(block1, block2) : nullptr;
        const bool fill1 = FillBlock(ancestor, ancestor_out, lock);
        const bool fill2 = FillBlock(block1, block1_out, lock);
        const bool fill3 = FillBlock(block2, block2_out, lock);
        return fill1 && fill2 && fill3;
    }
```

The current formulation fires off static analyser warnings for all static analysers checking the AutoSAR rules ("Guidelines for the use of the C++14 language in critical and safety-related systems") or MISRA C++.

AutoSAR ("Guidelines for the use of the C++14 language in critical and safety-related systems"):

> Rule M4-5-1 (required, implementation, automated): Expressions with type bool shall not be used as operands to built-in operators other than the assignment operator =, the logical operators &&, ||, !, the equality operators == and ! =, the unary & operator, and the
conditional operator.

ryanofsky commented:
Feel free to submit a PR. I'd also encourage writing documentation on how to run these static analysers with bitcoin, or maybe run them automatically on travis

MarcoFalke commented:
Oh, I didn't see this during review. I read the method as nothing is filled when the ancestor is not found. I.e. I read the `&` as `&&`. Is there any caller that depends on this edge case?

Absent any data, I'd slightly prefer the `&&` version.

ryanofsky commented:
Pretty sure nothing requires information about other blocks to be filled when a block isn't found in this case. It just seemed nicer for the API to return information than not return it, and trivial to implement, so I implemented it.

practicalswift commented:
Which one is preferred of these three? :)

Contender 1.

```diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index c8311b298..cd825f870 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -275,7 +275,7 @@ public:
         const CBlockIndex* block1 = LookupBlockIndex(block_hash1);
         const CBlockIndex* block2 = LookupBlockIndex(block_hash2);
         const CBlockIndex* ancestor = block1 && block2 ? LastCommonAncestor(block1, block2) : nullptr;
-        return FillBlock(ancestor, ancestor_out, lock) & FillBlock(block1, block1_out, lock) & FillBlock(block2, block2_out, lock);
+        return FillBlock(ancestor, ancestor_out, lock) && FillBlock(block1, block1_out, lock) && FillBlock(block2, block2_out, lock);
     }
     void findCoins(std::map<COutPoint, Coin>& coins) override { return FindCoins(m_node, coins); }
     double guessVerificationProgress(const uint256& block_hash) override
```

Contender 2.

```diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index c8311b298..ed2142884 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -275,7 +275,10 @@ public:
         const CBlockIndex* block1 = LookupBlockIndex(block_hash1);
         const CBlockIndex* block2 = LookupBlockIndex(block_hash2);
         const CBlockIndex* ancestor = block1 && block2 ? LastCommonAncestor(block1, block2) : nullptr;
-        return FillBlock(ancestor, ancestor_out, lock) & FillBlock(block1, block1_out, lock) & FillBlock(block2, block2_out, lock);
+        const bool ancestor_filled = FillBlock(ancestor, ancestor_out, lock);
+        const bool block1_filled = FillBlock(block1, block1_out, lock);
+        const bool block2_filled = FillBlock(block2, block2_out, lock);
+        return ancestor_filled && block1_filled && block2_filled;
     }
     void findCoins(std::map<COutPoint, Coin>& coins) override { return FindCoins(m_node, coins); }
     double guessVerificationProgress(const uint256& block_hash) override
```

Contender 3.

```diff
(empty diff)
```

MarcoFalke commented:
My vote:

1) 51%
2) 49%
3) 0%

ryanofsky commented:
I prefer the current code but again please open a separate PR, I don't think it matters very much

MarcoFalke commented:
In commit f6da44ccce : 

The following comment no longer applies and should be fixed up:

```cpp
            // Get required locks upfront. This avoids the GUI from getting
            // stuck if the core is holding the locks for a longer time - for
            // example, during a wallet rescan.
            //
```

MarcoFalke commented:
(somehow GitHub didn't submit my review comment in time ^)
